dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java

```diff
@@ -861,141 +858,180 @@ protected SolrQuery resolveToSolrQuery(Context context, DiscoverQuery discoveryQ
+                            name = facetQuery.substring(facetQuery.indexOf('[') + 1);
+                            name = name.substring(0, name.lastIndexOf(']')).replaceAll("TO", "-");
+                            filter = facetQuery.substring(facetQuery.indexOf('['));
+                            filter = filter.substring(0, filter.lastIndexOf(']') + 1);
+                        }
+                        Integer count = sortedFacetQueries.get(facetQuery);
+
+                        //No need to show empty years
+                        if (0 < count) {
+                            result.addFacetResult(facetField,
+                                                  new DiscoverResult.FacetResult(filter, name, null, name, count,
+                                                                                 DiscoveryConfigurationParameters
+                                                                                     .TYPE_DATE));
+                        }
                     }
+                }
 
-                    Integer count = sortedFacetQueries.get(facetQuery);
-
-                    //No need to show empty years
-                    if (0 < count) {
-                        result.addFacetResult(facetField,
-                                              new DiscoverResult.FacetResult(filter, name, null, name, count,
-                                                                             DiscoveryConfigurationParameters
-                                                                                 .TYPE_DATE));
+                if (solrQueryResponse.getSpellCheckResponse() != null && !zombieFound) {
+                    String recommendedQuery = solrQueryResponse.getSpellCheckResponse().getCollatedResult();
+                    if (StringUtils.isNotBlank(recommendedQuery)) {
+                        result.setSpellCheckQuery(recommendedQuery);
                     }
                 }
             }
-
-            if (solrQueryResponse.getSpellCheckResponse() != null) {
-                String recommendedQuery = solrQueryResponse.getSpellCheckResponse().getCollatedResult();
-                if (StringUtils.isNotBlank(recommendedQuery)) {
-                    result.setSpellCheckQuery(recommendedQuery);
-                }
+            // If any stale entries are found in the current page of results,
+            // we remove those stale entries and rerun the same query again.
+            // Otherwise, the query is valid and the results are returned.
+            if (zombieDocs.size() != 0) {
+                log.info("Cleaning " + zombieDocs.size() + " stale objects from Discovery Index");
+                solrSearchCore.getSolr().deleteById(zombieDocs);
+                solrSearchCore.getSolr().commit();
+            } else {
+                valid = true;
             }
+        } while (!valid && executionCount <= maxAttempts);
+
+        if (!valid && executionCount == maxAttempts) {
```

Comment:
If `maxAttempts` ends up set to zero, this `if` will never be triggered (as executionCount has a minimum value of 1).